README.md
##

Note: This is a copy of the [original distribution](http://sourceforge.net/projects/tkpacman/)

All that has changed here is I hacked around with it to make it run on my windows MSY2 environment and added reconciliation after applying changes so that any changes that failed to be applied are still selected.

To run on MSYS2 I made sure I had tcl and tk installed and then changed in to the directory with this in and ran `tclsh main.tcl` and set the options appropriately.  I also had to decrease the font size a bit to make the rows readable.

If you find this, feel free to do with it what you will, I needed something and this was my way of getting there.  I doubt I'll come back to it :-)

# Original Readme

tkPacman is a graphical user interface for pacman, the package
manager for Arch Linux.

Copyright © 2013-2020 Willem Herremans

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

A copy of the  GNU General Public License is included in the 'license'
subdirectory of this software package.

The home page of tkPacman can be found at:

http://sourceforge.net/projects/tkpacman/

There you can report bugs, request new features and get support.

tkPacman is completely written in Tcl/Tk. There is no need to compile
or build anything. The best way to install it is to use the AUR package
at:

https://aur.archlinux.org/packages/tkpacman/


CHANGE_LOG.txt
Version 1.9.1 (2020-01-15)
  - The filter in the file selection box called from "Install package
    from local file" has been modified from "*.pkg.tar.xz" to
    "*.pkg.tar.*". As a result, both packages with file extensions
    ".pkg.tar.xz" and ".pkg.tar.zst" are shown in the file selection
    box.
Version 1.9.0 (2018-07-08)
  - Possibility to call "pacman-optimize" has been removed because
    pacman also does not provide it anymore.
  - "gksu" has been removed from option "runasroot" because "gksu" has
    been removed from the Arch linux repositories
  - "vte" has been replaced with "vte-2.91" in option "terminal".
Version 1.8.0 (2017-10-22)
  - Corrections to the predefined templates for the "terminal" option:
      - roxterm was removed because it currently does not work wel
        with tkPacman
      - The template for konsole was corrected
      - A template for qterminal was added.
      - The help text for this option was corrected accordingly
  - The runasroot option "sudo" now also works in KDE without having to
    modify the sudo configuration in an awkward way. The help text was
    modified accordingly.
  - The AUR PKGBULD file now uses a Makefile. I thank Alad Wenter for
    showing me how to set up a Makefile.

Version 1.7.0 (2016-09-12)
  - The option "fontincrement" has been added. It saves the font size
    used by tkPacman.
  - Correction in the help text: the hyperlink to "Partial upgrades
    are unsupported" has been corrected.

Version 1.6.1 (2015-10-13)
  - The button "Check Upgrades" was renamed to "List Upgrades".
  - tkPacman no longer calls the pacman script "checkupdates" but
    provides the same functionality by setting up a temporary copy of the
    package database at /tmp/tkpacman, where /tmp/tkpacman/local is a
    symbolic link to /var/lib/pacman/local, and by calling:
        fakeroot -- pacman --sync --refresh --dbpath /tmp/tkpacman
        pacman --query --upgrades --dbpath /tmp/tkpacman
    This gives the user more feedback on the progress than 'checkupdates'
    which hides all feedback on the progress from pacman.
  - The parsing of /etc/pacman.conf was completely rewritten. Up to now
    it is used by tkPacman to get the list of sofware repositories and
    the value of DBPath (default /var/lib/pacman).

Version 1.6.0 (2015-09-27)
  - The button "Resync Database", which called 'pacman --sync --refresh',
    has been replaced with a button "Check Upgrades", which calls
    'checkupdates'. I would like to thank Alad Wenter for pointing out
    to me that 'pacman --sync --refresh' is unsafe in the sense that it
    increases the risk for a "partial upgrade". See also the pacman wiki
    on the Archlinux website. The help text of tkPacman has been updated
    accordingly.
  - For the time being, the call to 'checkupdates' is preceded by
    setting LANG=en_US.UTF-8. This is necessary now to work around a bug
    in 'checkupdates' for the Dutch locale. See
    https://bugs.archlinux.org/task/46405
  - The template for the terminal 'lxterminal' has been removed. I
    have noticed that if an lxterminal is already running outside
    tkPacman, lxterminal returns control to tkPacman immediately, even
    before the command '%c' has started. Therefore, I can no longer
    recommend it for use with tkPacman.
  - The template for the terminal 'roxterm' was modified. The option
    '--separate' was added to it. This was necessary to prevent it form
    reusing the process of another already running roxterm.
  - A new menu item was added to the Pacman menu:
    "Optimize pacman database". It calls 'pacman-optimize'.
  - The Dutch translation of "repository" was changed from "depot" to
    "archief".


Version 1.5.0 (2014-04-13)
  - The action 'Clean cache' in menu 'Pacman' now offers the choice
    to clean up all packages or only the ones that are no longer
    installed.
  - The statement 'Defaults env_keep += "HOME"' which was recommended
    in the help text for configuring 'sudo' has some undesired effects.
    The help text has been changed accordingly. If you are using
    tkPacman in KDE, it is probably better to use 'kdesu' instead
    of 'sudo' and to remove the offending statement form /etc/sudoers
    by calling 'visudo'.

Version 1.4.0 (2014-01-03)
  - tkPacman now also remembers packages marked for installation or
    removal across mode switches (from 'Available packages' to
    'Installed packages' or vice versa).
  - It is possible to mark packages for removal if there are already
    packages marked for installation or vice versa.
  - When the 'Apply changes' button is pressed, first the packages
    marked for removal are removed and then the packages marked for
    installation are installed.

Version 1.3.1 (2013-11-29)
  - Ticket number 5. tkPacman crashes on startup in most non-English
    locales. Solved in procedure parseFirst in main.tcl.

Version 1.3.0 (2013-11-2)
  - Important error messages are not printed to stderr, but are stored
    in internal memory and can be viewed by pressing the 'Errors' button
    on the toolbar. Additionally, the 'Errors' button's icon changes
    from gray to red when new errors have occurred. This has been done
    to help new users who propably don't know what is happening when
    the 'terminal' or 'runasroot' options are calling uninstalled
    programs.

Version 1.2.0 (2013-10-30)
  - tkPacman now remembers packages marked for installation or
    removal across filter changes. Only when the user changes from
    'available packages' to 'installed packages' or vice versa,

askpass/askpass.tcl
#!/usr/bin/tclsh
# This script is meant to be called by 'sudo --askpass'
# sudo calls it with 1 argument containing text as:
#   [sudo] password for wim:
# This helper must be "registered" by tkPacman as sudo helper
# by setting the environment variable SUDO_ASKPASS to this script file.

package require Tk
package require msgcat
namespace import msgcat::mc

set installDir [file dirname [file normalize [info script]]]
set languageDir [file join $installDir msgs]
msgcat::mcload $languageDir

# Set all backgrounds equal
# Without these adjustments, there are some ugly background colour
# differences in the KDE desktop. The toplevel windows take their
# background colour from KDE and the other widgets take it from
# the Tk theme.
set themeBackground [ttk::style lookup TFrame -background]
{.} configure -background $themeBackground

proc appendBindTag {widget newTag} {
    set tags [bindtags $widget]
    lappend tags $newTag
    bindtags $widget $tags
    return
}

proc onDestroy {} {
    bind topAskpass <Destroy> {}
    exit
}

proc onOK {} {
    global password
    chan puts stdout $password
    bind topAskpass <Destroy> {}
    exit
}

wm title {.} [mc askPassword]

set password {}
set lblHelp [ttk::label .lblHelp -text [mc sudoHelp [lindex $argv 0]]]
set fPassword [ttk::frame .fPassword]
set lblPassword [ttk::label ${fPassword}.lblPassword -text [mc password]]
set entPassword [ttk::entry ${fPassword}.entPassword -show {*} \
    -textvariable password]
set fButtons [ttk::frame .fButtons]
set btnCancel [ttk::button ${fButtons}.btnCancel -text [mc btnCancel] \
    -command [list destroy .] -takefocus 0]
set btnOK [ttk::button ${fButtons}.btnOK -text [mc btnOK] \
    -command onOK -takefocus 0]
bind . <KeyPress-Return> onOK
bind . <KeyPress-Escape> [list destroy .]
# bind <Destroy> for toplevel only
appendBindTag . topAskpass
bind topAskpass <Destroy> [list onDestroy]
focus $entPassword
pack $lblPassword -side left
pack $entPassword -side left -expand 1 -fill x
pack $btnOK -side right
pack $btnCancel -side right
pack $lblHelp -side top -pady 10 -padx 10
pack $fPassword -side top -pady 10 -padx 10 -fill x
pack $fButtons -side top -pady 10 -padx 10 -fill x


options.tcl
# options.tcl

namespace eval pacmanOptions {
    # This namespace encapsulates all variable and procedures related to
    # the pacman options as stored in /etc/pacman.conf. pacmanDict is a
    # dictionary of dictionaries. The outer dictionary uses the section
    # names (options, core, extra, community ...) as keys and the section
    # bodies as values. Each section body is itself a dictionary with
    # what comes before the '=' as key, and the rest of the line as
    # value. If there is no '=' in the line, the empty string {} is used
    # as value.
    variable pacmanDict {}
    namespace ensemble create -subcommands {initOptions getOption getRepolist}
}

namespace eval pacmanOptions {
    # This procedure reads /etc/pacman.conf and populates pacmanDict.
    proc initOptions {} {
        variable pacmanDict
        if {![catch {open "/etc/pacman.conf" r} confchan]} then {
            set pacmanConf [split [chan read $confchan] "\n"]
            chan close $confchan
            # remove all comments and empty lines
            set index 0
            foreach line $pacmanConf {
                # remove all empty space before and after line
                set line [string trim $line]
                if {[string length $line] == 0} then {
                    # empty line, remove it
                    set pacmanConf [lreplace $pacmanConf $index $index]
                    incr index -1
                } else {
                    set hash [string first "#" $line 0]
                    if {$hash == 0} then {
                        # first char of line is "#" --> whole line is comment,
                        # remove it
                        set pacmanConf [lreplace $pacmanConf $index $index]
                        incr index -1
                    } elseif {$hash > 0} then {
                        # line contains comment, remove comment
                        set line [string replace $line $hash end ""]
                        set pacmanConf [lreplace $pacmanConf $index $index $line]
                    } else {
                        # no comments and not empty, but line may be trimmed -->
                        # it must be replaced nevertheless.
                        set pacmanConf [lreplace $pacmanConf $index $index $line]
                    }
                }
                incr index
            }
        } else {
            set pacmanConf {}
        }
        set pacmanDict {}
        set section {}
        set body {}
        foreach line $pacmanConf {
            if {[string index $line 0] eq {[}} then {
                # new section --> append accumulated lines
                # to previous section
                if {$section ne {}} then {
                    # There is a previous section
                    dict append pacmanDict $section $body
                    set body {}
                }
                set end [string first {]} $line]
                set section [string range $line 1 ${end}-1]
            } else {
                # section continues. Append line to body
                set line [split $line {=}]
                set key [string trim [lindex $line 0]]
                if {[llength $line] == 1} then {
                    set value {}
                } else {
                    set value [string trim [lindex $line 1]]
                }
                dict append body $key $value
            }
        }
        if {$section ne {}} then {
            dict append pacmanDict $section $body
        }
        # puts $pacmanDict
        return
    }
}

namespace eval pacmanOptions {
    proc getOption {section option} {
        variable pacmanDict
        if {[dict exists $pacmanDict $section $option]} then {
            set value [dict get $pacmanDict $section $option]
        } else {
            set value {}
        }
        return $value
    }
}

namespace eval pacmanOptions {
    proc getRepolist {} {
        variable pacmanDict
        set repolist {}
        foreach key [dict keys $pacmanDict] {
            if {$key ne {options}} then {
                lappend repolist $key
            }
        }
        return $repolist
    }
}


##
# namespace tkpOptions
#   This namespace encapsulates all variables and procedures
#   for handling the options of tkPacman.

namespace eval tkpOptions {
    variable optionStructure [list \
        general [list browser terminal runasroot tmpdir allerrors fontincrement] \
        geometry [list main text import] \
        fileselection [list lastdir]]
    variable optionDict
    variable optionArray
    variable window
    variable terminalList
    variable runasrootList
    set optionDict {}
    dict for {type list} $tkpOptions::optionStructure {
        set subDict {}
        foreach option $list {
            dict append subDict $option {}
        }
        dict append optionDict $type $subDict
    }
    set window {}
    # WHE : terminalList updated 22-10-2017
    # roxterm does not work very well anymore --> removed
    # konsole template corrected
    # qterminal added
    set terminalList [list \
        {xfce4-terminal --disable-server --title=%t --command=%c} \
        {konsole --nofork -p tabtitle=%t -e %c} \
        {mate-terminal --disable-factory --title=%t --command=%c} \
        {vte-2.91 --name=%t --command=%c} \
        {qterminal --execute %c} \
        {xterm -title %t -e %c} \
        {C:\\msys64\\msys2_shell.cmd -c %c}]
    set runasrootList [list \
        {%terminal(echo "%p" ; su --command="%p" ; read -p "%close")} \
        {sudo --askpass %terminal(echo "%p" ; %p ; read -p "%close")} \
        {kdesu %terminal(echo "%p" ; %p ; read -p "%close")}]
    namespace ensemble create -subcommands [list getDefault \
        getOption initOptions saveOptions setOption showWindow]
}

namespace eval tkpOptions {
    proc initOptions {} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window

        set filename [file normalize "~/.tkpacman"]
        if {[file exists $filename]} then {
            set chan [open $filename r]
            set fileDict [chan read -nonewline $chan]
            chan close $chan
        } else {
            set fileDict {}
        }
        dict for {type list} $tkpOptions::optionStructure {
            foreach option $list {
                if {[dict exists $fileDict $type $option]} then {
                    dict set optionDict $type $option \
                        [dict get $fileDict $type $option]
                } else {
                    dict set optionDict $type $option \
                        [getDefault $type $option]
                }
            }
        }
        saveOptions
        return
    }
}


namespace eval tkpOptions {
    proc saveOptions {} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window

        set filename [file normalize "~/.tkpacman"]
        set chan [open $filename w]
        chan puts $chan $optionDict
        chan close $chan
        return
    }
}

namespace eval tkpOptions {
    proc getDefault {type option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        switch $type {
            general {
                set value [getGeneralDefault $option]
            }
            geometry {
                set value [getGeometryDefault $option]
            }
            fileselection {
                set value [getFileselectDefault $option]
            }
            default {
                set value {}
            }
        }
        return $value
    }
}

namespace eval tkpOptions {
    proc getGeneralDefault {option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        variable terminalList
        variable runasrootList
        variable ::config::defaultBrowser

        switch -- $option {
            "browser" {
                set value {/usr/bin/xdg-open}
            }
            "terminal" {
                set value [lindex $terminalList 0]
                foreach terminal $terminalList {
                    set termexec [lindex $terminal 0]
                    if {![catch {exec which $termexec} msg]} then {
                        set value $terminal
                        break
                    }
                }
            }
            "runasroot" {
                set value [lindex $runasrootList 0]
            }
            "tmpdir" {
                set value {/tmp}
            }
            "allerrors" {
                set value 0
            }
            "fontincrement" {
                set value 0
            }
            default {
                set value {}
            }
        }
        return $value
    }
}

namespace eval tkpOptions {
    proc getGeometryDefault {option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        switch $option {
            main {
                set value {750 450}
            }
            text {
                set value {600 400}
            }
            import {
                set value {600 400}
            }
            default {
                set value {600 400}
            }
        }
        return $value
    }
}

namespace eval tkpOptions {
    proc getFileselectDefault {option} {
        if {$option eq "lastdir"} then {
            set value [file normalize ~]
        } else {
            set value {}
        }
        return $value
    }
}

namespace eval tkpOptions {
    proc getOption {type option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        return [dict get $optionDict $type $option]
    }
}

namespace eval tkpOptions {
    proc setOption {type option newValue} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        dict set optionDict $type $option $newValue
        return
    }
}

namespace eval tkpOptions {
    proc showWindow {parent} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        if {$window eq {}} then {
            foreach type {general} {
                foreach option [dict get $optionStructure $type] {
                    set optionArray($type,$option) [dict get $optionDict $type $option]
                }
            }
            set window [toplevel [appendToPath $parent options]]
            wm transient $window $parent
            wm title $window [mc optionsTitle]
            # set noteBook [ttk::notebook ${window}.nb -takefocus 0]
            set fGeneral [setupTab general ${window}.general]
            #addNotebookTab $noteBook $fGeneral optTabGeneral
            #ttk::notebook::enableTraversal $noteBook
            #pack $noteBook -side top -expand 1 -fill both
            pack $fGeneral -side top -expand 1 -fill both -padx 5 -pady 5
            set frmButtons [ttk::frame ${window}.frmbtns]
            set btnLegend [defineButton ${frmButtons}.btnlegend $window \
                btnLegend tkpOptions::onLegend]
            set btnOK [defineButton ${frmButtons}.btnok $window btnOK \
                tkpOptions::onOK]
            set btnCancel [defineButton ${frmButtons}.btncancel $window \
                btnCancel [list destroy $window]]
            pack $btnLegend -side right
            pack $btnCancel -side right
            pack $btnOK -side right
            pack $frmButtons -side top -fill x -pady 5 -padx 5
            bindToplevelOnly $window <Destroy> tkpOptions::onDestroyWindow
            #bind $window <KeyPress-Down> {focus [tk_focusNext [focus]]}
            #bind $window <KeyPress-Up> {focus [tk_focusPrev [focus]]}
            bind $window <KeyPress-Return> [list tkpOptions::onOK]
            bind $window <KeyPress-Escape> [list destroy $window]
        } else {
            wm deiconify $window
            raise $window
            focus $window
        }
        return
    }
}

namespace eval tkpOptions {
    proc setupTab {tab pathname} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        set entrywidth 40
        set frm [ttk::frame $pathname]
        set idx 0
        foreach option [dict get $tkpOptions::optionStructure $tab] {
            set lbl [ttk::label ${frm}.lb$idx -text $option \
                -compound right -image ::img::empty_5_9]
            grid $lbl -column 0 -row $idx -sticky w
            if {$option eq {runasroot}} then {
                set control [ttk::combobox ${frm}.con$idx \
                    -textvariable tkpOptions::optionArray($tab,$option) \
                    -values $tkpOptions::runasrootList]
            } elseif {$option eq {terminal}} then {
                set control [ttk::combobox ${frm}.con$idx \
                    -textvariable tkpOptions::optionArray($tab,$option) \
                    -values $tkpOptions::terminalList]
            } elseif {$option eq {allerrors}} then {
                set control [ttk::checkbutton ${frm}.con$idx \
                    -variable tkpOptions::optionArray($tab,$option) \
                    -onvalue 1 -offvalue 0]
            } elseif {$option eq {fontincrement}} then {
                set control [entry ${frm}.con$idx \
                    -textvariable tkpOptions::optionArray($tab,$option)]
                $control configure -state readonly
            } else {
                set control [entry ${frm}.con$idx \
                    -textvariable tkpOptions::optionArray($tab,$option)]
                $control configure -width $entrywidth
            }
            bind $control <FocusIn> \
                [list $lbl configure -compound right -image ::img::arrow_right]
            bind $control <FocusOut> \
                [list $lbl configure -compound right -image ::img::empty_5_9]
            if {$idx == 0} then {
                focus $control
            }
            if {$option in {browser tmpdir}} then {
                set btnSelect [ttk::button ${frm}.sel$idx \
                    -image ::img::arrow_down \
                    -takefocus 0 \
                    -command [list tkpOptions::onSelect $tab $option]]
                bind $control <KeyPress-Down> [list $btnSelect invoke]
                bind $control <Alt-KeyPress-s> [list $btnSelect invoke]
                grid $control -column 1 -row $idx -sticky wens
                grid $btnSelect -column 2 -row $idx -sticky wens
                # $control configure -state {readonly}
            } else {
                grid $control -column 1 -columnspan 2 -row $idx \
                    -sticky wens
            }
            set btnExpand [ttk::button ${frm}.exp$idx \
                -image ::img::expand \
                -takefocus 0 \
                -command [list tkpOptions::onExpand $tab $option]]
            bind $control <Alt-KeyPress-x> [list $btnExpand invoke]
            grid $btnExpand -column 3 -row $idx -sticky wens
            set btnDefault [ttk::button ${frm}.def$idx \
                -image ::img::reset \
                -takefocus 0 \
                -command [list tkpOptions::onDefault $tab $option]]
            bind $control <Alt-KeyPress-r> \
                [list tkpOptions::onDefault $tab $option]
            grid $btnDefault -column 4 -row $idx -sticky wens
            set btnHelp [ttk::button ${frm}.help$idx \
                -image ::img::help \
                -takefocus 0 \
                -command [list tkpOptions::onHelp $tab $option]]
            bind $control <Alt-KeyPress-h> [list $btnHelp invoke]
            grid $btnHelp -column 5 -row $idx -sticky wens
            incr idx
        }
        grid columnconfigure $frm 1 -weight 1
        grid anchor $frm center
        return $frm
    }
}

namespace eval tkpOptions {
    proc onOK {} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        foreach type {general} {
            foreach option [dict get $tkpOptions::optionStructure $type] {
                dict set optionDict $type $option $optionArray($type,$option)
            }
        }
        saveOptions
        destroy $window
        return
    }
}

namespace eval tkpOptions {
    proc onDestroyWindow {} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        array unset optionArray
        set window {}
        return
    }
}

namespace eval tkpOptions {
    proc onExpand {type option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        set textEdit [TextEdit new $window $option \
            $optionArray($type,$option) 0]
        if {$option in {browser terminal}} then {
            $textEdit addMenuItem [mc optPasteFilename] command {
                %T insert insert [fileSelectionFunction \
                    -parent [winfo toplevel %T]]
            }
        }
        if {[$textEdit wait result]} then {
            set optionArray($type,$option) $result
        }
        return
    }
}

namespace eval tkpOptions {
    proc onDefault {type option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        set optionArray($type,$option) [getDefault $type $option]
        return
    }
}

namespace eval tkpOptions {
    proc onHelp {type option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        set textEdit [TextEdit new $window [mc optHelpTitle $option] \
                [mc optHelp_${option}] 1]
        return
    }
}

namespace eval tkpOptions {
    proc onLegend {} {
        variable window
        set llist {}
        foreach icon {arrow_down expand reset help} {
            lappend llist [list ::img::$icon [mc opt_$icon]]
        }
        set legend [Legend new $window [mc fsLegend] $llist]
        return
    }
}

namespace eval tkpOptions {
    proc onSelect {type option} {
        variable optionStructure
        variable optionDict
        variable optionArray
        variable window
        variable terminalList
        global env
        global tcl_platform
        if {$option in {browser}} then {
            set initialdir {/usr/bin}
            set fs [FileSelection new -parent $window \
                -directory $initialdir \
                -title [mc optSelect$option]]
            set filename [$fs wait]
            if {$filename ne {}} then {
                set optionArray($type,$option) $filename
            }
        } elseif {$option in {tmpdir}} then {
            set initialdir {/tmp}
            set fs [FileSelection new -parent $window \
                -directory $initialdir \
                -title [mc optSelect$option] \
                -dironly 1]
            set dirname [$fs wait]
            if {$dirname ne {}} then {
                set optionArray($type,$option) $dirname
            }
        }
        return
    }
}
